File: functions/makeEFieldMap.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.constants as cons
import json

#3次元プロットするためのモジュール
from mpl_toolkits.mplot3d import Axes3D
#行列N×N
N = 10
#演算範囲
lim = 0.5
#刻み幅
dr = lim*2/N
#真空中の誘電率
e = 8.854e-12
V = 168 #[V/m]
def m2Dex(m):

    return int((m + lim)/dr)


print("dr",lim*2/N)
#とりあえず二次元
phimap = [[[0 for k in range(N)] for j in range(N)] for i in range(N)]
flagmap = [[[False for k in range(N)] for i in range(N)] for j in range(N)]
#とりあえず、y=0.1,-0.1,-0.1<x<0.1のとこに電荷置く
for z in np.arange(-0.2,0.2,dr):
    for y in np.arange(-0.2,0.2,dr):
        xDex = m2Dex(0.2)
        yDex = m2Dex(y)
        zDex = m2Dex(z)
        phimap[xDex][yDex][zDex] = V/2
        flagmap[xDex][yDex][zDex] = True
        yDex = m2Dex(-0.2)
        phimap[xDex][yDex][zDex] = -V/2
        flagmap[xDex][yDex][zDex] = True
#print(rhomap)
count = 0

MaxErr = 1
Conv = 10**(-6)
MaxPhi = 1
while MaxErr>Conv:
    #print(phimap)
    MaxErr = CurErr = 0.0
    count = 1 + count

    for x in range(1,N-1):
        for y in range(1,N-1):
            for z in range(1,N-1):
                if flagmap[x][y][z]:
                    continue
                Prev_phi = phimap[x][y][z]
                tmp =1/6 * (phimap[x+1][y][z]+phimap[x-1][y][z]+phimap[x][y+1][z]+phimap[x][y-1][z]+phimap[x][y][z+1]+phimap[x][y][z-1])
                if MaxPhi < abs(tmp):
                    MaxPhi = abs(tmp)
                phimap[x][y][z] = tmp
                CurErr = (abs(phimap[x][y][z] - Prev_phi))/MaxPhi
                if MaxErr < CurErr:
                     MaxErr = CurErr
    if count % 1 == 0:
        print(count,MaxErr)
with open('./data/phimap.json','w') as fw:
    json.dump(phimap,fw,indent=4)
    fw.close()
np.save("phimap",phimap)

File: functions/makeEfieldMap.c
#include<stdio.h>
#include<stdbool.h>
#include<stdlib.h>
#include<math.h>
#define N 200
#define LIM 0.5
#define e 8.854e-12
#define V 16800
#define dr LIM*2.0/N
int m2Dex(float m){
  //printf("%lf\n dr = %lf\n",m+LIM,dr);
  float tmp1 = dr;
  float tmp2 = m+LIM;
  float tmp3 = tmp2/tmp1;
  //printf("1 = %lf\n 2  = %lf\n 3 = %lf\n",tmp1,tmp2,tmp3);
  //printf("int = %d\n",(int)round(tmp3));
  return ((int)round(tmp3));
}

double ***malloc_3dim_phimap(int n1, int n2, int n3) {
  int i, j;
  double ***array;
  array = (double ***)malloc(n1 * sizeof(double **));
  for (i = 0; i < n1; i++) {
    array[i] = (double **)malloc(n2 * sizeof(double *));
    for (j = 0; j < n2; j++)
      array[i][j] = (double*)malloc(n3 * sizeof(double));
  }
  return array;
}
bool ***malloc_3dim_flagmap(int n1, int n2, int n3) {
  int i, j;
  bool ***array;
  array = (bool ***)malloc(n1 * sizeof(bool **));
  for (i = 0; i < n1; i++) {
    array[i] = (bool **)malloc(n2 * sizeof(bool *));
    for (j = 0; j < n2; j++)
      array[i][j] = (bool*)malloc(n3 * sizeof(bool));
  }
  return array;
}
/* 3次元配列の開放 */
void free_3dim_array(void ***array, int n1, int n2, int n3) {
  int i, j;
  for (i = 0; i < n1; i++) {
    for (j = 0; j < n2; j++)
      free(array[i][j]);
    free(array[i]);
  }
  free(array);
}

void print_3dim_array(double ***array, int n1, int n2, int n3) {
  int i, j, k;
  for (i = 1; i < n1-1; i++) {
    for (j = 1; j < n2-1; j++) {
      for (k = 1; k < n3-1; k++)
        printf("%lf ", array[i][j][k]);
      printf("\n");
    }
    printf("\n");
  }
}



int main(){
  printf("start\n");
  double ***phimap = malloc_3dim_phimap(N, N, N);
  bool ***flagmap = malloc_3dim_flagmap(N, N, N);
  for(int i = 0 ; i < N; i++) {
    for(int j = 0 ; j < N; j++) {
      for(int k = 0 ; k < N; k++) {
        phimap[i][j][k] = 0;
        flagmap[i][j][k] = false;
      }
    }
  }
  for(float z = -0.2;z<0.2;z = z + dr){
    for(float y = -0.2;y<0.2;y = y + dr){
      int xDex = m2Dex(0.2);
      int yDex = m2Dex(y);
      int zDex = m2Dex(z);
      //printf("%d %d %d \n",xDex,yDex,zDex);
      phimap[xDex][yDex][zDex] = V;
      flagmap[xDex][yDex][zDex] = true;
      yDex = m2Dex(-0.2);
      phimap[xDex][yDex][zDex] = 0;
      flagmap[xDex][yDex][zDex] = true;
    }
  }
  //print_3dim_array(phimap,N,N,N);
  int count = 0;
  double maxerr = 1;
  double conv = 10e-6;
  double curerr;
  double prev_phi;
  double tmp;
  double maxphi = 1;
  printf("loop start\n conv = %lf\nmaxerr = %lf\n",conv,maxerr);
  while(maxerr>conv){
    //printf("test\n");
    maxerr = 0;
    curerr = 0;
    count++;
    for(int i = 1;i<N-1;i++){
      for(int j = 1;j<N-1;j++){
        for(int k = 1;k<N-1;k++){
          //print_3dim_array(phimap,N,N,N);
          if(flagmap[i][j][k]){
            continue;
          }
          prev_phi = phimap[i][j][k];
          tmp = (phimap[i+1][j][k] + phimap[i-1][j][k] + phimap[i][j+1][k] + phimap[i][j-1][k] + phimap[i][j][k+1] + phimap[i][j][k-1])/6;
          //printf("tmp = %lf\n",tmp);
          if(maxphi<abs(tmp)){
            maxphi = abs(tmp);
          }
          phimap[i][j][k] = tmp;
          curerr = (abs(phimap[i][j][k] - prev_phi))/maxphi;
          //printf("curerr = %lf\n phimap = %lf\n prev_phi = %lf\n maxphi  = %lf\n",curerr,phimap[i][j][k],prev_phi,maxphi);
          if(maxerr < curerr){
            maxerr = curerr;
          }
        }
      }
    }
    //printf("maxerr = %lf\nconv = %lf\n",maxerr,conv);
    if(count % 100 == 0){
      printf("count = %d\n maxerr = %lf\n",count,maxerr);
    }
  }
  printf("%d\n",count);
  free_3dim_array((void***)phimap, N, N, N);
  free_3dim_array((void***)flagmap, N, N, N);

  printf("end\n");

  return 0;
}
